src/dialog/unix/SDL_dialog.c

```diff
@@ -0,0 +1,394 @@
+/*
+  Simple DirectMedia Layer
+  Copyright (C) 1997-2023 Sam Lantinga <slouken@libsdl.org>
+
+  This software is provided 'as-is', without any express or implied
+  warranty.  In no event will the authors be held liable for any damages
+  arising from the use of this software.
+
+  Permission is granted to anyone to use this software for any purpose,
+  including commercial applications, and to alter it and redistribute it
+  freely, subject to the following restrictions:
+
+  1. The origin of this software must not be misrepresented; you must not
+     claim that you wrote the original software. If you use this software
+     in a product, an acknowledgment in the product documentation would be
+     appreciated but is not required.
+  2. Altered source versions must be plainly marked as such, and must not be
+     misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+*/
+#include "SDL_internal.h"
+
+/* TODO: Macro? */
+
+#include <errno.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <unistd.h>
+
+typedef enum
+{
+    ZENITY_MULTIPLE = 0x1,
+    ZENITY_DIRECTORY = 0x2,
+    ZENITY_SAVE = 0x4
+} zenityFlags;
+
+typedef struct
+{
+    SDL_DialogFileCallback callback;
+    void* userdata;
+    const char* filename;
+    const SDL_DialogFileFilter *filters;
+    Uint32 flags;
+} zenityArgs;
+
+/* TODO: Zenity survives termination of the parent */
+
+static void run_zenity(zenityArgs* arg_struct)
```

Comment:
If you still need and would be interested in discussing some here, I can assist in making this use portals appropriately.

Reply:
I must confess that I have a hard time understanding the inner workings of @quyykk's solution, so help would be appreciated there.

@slouken Would it be preferable to implement only portals on Linux, or also keep Zenity as a fallback?

Reply:
> @slouken Would it be preferable to implement only portals on Linux, or also keep Zenity as a fallback?

No, we'll want to keep zenity as a fallback.